src/server/pmix_server_ops.c

```diff
@@ -2056,47 +2221,68 @@ pmix_status_t pmix_server_register_events(pmix_peer_t *peer,
+        if (PMIX_SUCCESS == (rc = pmix_host_server.register_events(scd->codes, scd->ncodes, scd->info, scd->ninfo, regevopcbfunc, scd))) {
+            /* the host will call us back when completed */
             pmix_output_verbose(2, pmix_server_globals.event_output,
-                                 "server register events: host server reg events returned rc =%d", rc);
-            if (NULL != scd->codes) {
-                free(scd->codes);
-            }
-            if (NULL != scd->info) {
-                PMIX_INFO_FREE(scd->info, scd->ninfo);
+                                 "server register events: host server processing event registration");
+            if (NULL != affected) {
+                free(affected);
             }
+            return rc;
+        } else if (PMIX_OPERATION_SUCCEEDED == rc) {
+            /* we need to check cached notifications, but we want to ensure
+             * that occurs _after_ the client returns from registering the
+             * event handler in case the event is flagged for do_not_cache.
+             * Setup an event to fire after we return as that means it will
+             * occur after we send the registration response back to the client,
+             * thus guaranteeing that the client will get their registration
+             * callback prior to delivery of an event notification */
+            PMIX_RETAIN(peer);
+            scd->peer = peer;
+            scd->procs = affected;
+            scd->nprocs = naffected;
+            scd->opcbfunc = NULL;
+            scd->cbdata = NULL;
+            PMIX_THREADSHIFT(scd, _check_cached_events);
+            return rc;
+        } else {
+            /* host returned a genuine error and won't be calling the callback function */
+            pmix_output_verbose(2, pmix_server_globals.event_output,
+                                 "server register events: host server reg events returned rc =%d", rc);
             PMIX_RELEASE(scd);
+            goto cleanup;
         }
     } else {
         rc = PMIX_OPERATION_SUCCEEDED;
+        /* we need to check cached notifications, but we want to ensure
+         * that occurs _after_ the client returns from registering the
+         * event handler in case the event is flagged for do_not_cache.
+         * Setup an event to fire after we return as that means it will
+         * occur after we send the registration response back to the client,
+         * thus guaranteeing that the client will get their registration
+         * callback prior to delivery of an event notification */
+        scd = PMIX_NEW(pmix_setup_caddy_t);
+        PMIX_RETAIN(peer);
+        scd->peer = peer;
+        scd->codes = codes;
+        scd->ncodes = ncodes;
+        scd->procs = affected;
+        scd->nprocs = naffected;
+        scd->opcbfunc = NULL;
+        scd->cbdata = NULL;
+        PMIX_THREADSHIFT(scd, _check_cached_events);
```

Comment:
The `DEBUGGER-RELEASE` case goes here since it's not an environmental event. So here we thread shift the `_check_cached_events` and return `PMIX_OPERATION_SUCCEEDED`. Does the return of this function then trigger the message back to the client to acknowledge the event (and thus trigger the client side registration callback)? I'm just not seeing where that acknowledgment is triggered - even with the old code.

Reply:
Yes - returning operation_succeeded to the pmix_server/switchyard causes the switchyard to return "success" to the client. Since the switchyard is operating inside an event, the message to the client gets queued prior to the threadshift being executed. Thus, the client gets the registration callback followed by the results of ```_check_cached_events```

Reply:
Perfect. Thanks!